text/0255-sensitive-data-redaction.md

```diff
@@ -0,0 +1,397 @@
+
+### Types of sensitive data
+
+There are different kinds of “sensitivity” that may apply to data. The ones most
+relevant in this proposal are “security” and “privacy”. They may overlap but we
+distinguish them as follows:
+
+- security-relevant sensitive data: any information that when exposed
+  [weakens the overall security of the device/system](https://en.wikipedia.org/wiki/Vulnerability_(computing)).
+- privacy-relevant sensitive data: any information that when exposed
+  [can adversely affect the privacy or welfare of an individual](https://en.wikipedia.org/wiki/Information_sensitivity).
+
+Note, that there are other kinds of sensitive data (business information,
+classified information), which are not covered extensively by this proposal.
+
+### Level of Sensitivity
+
+The level of sensitivity of an information can also be different and that
+sensitivity can be contextual, e.g.
+
+- The password of a user without privileges is less sensitive than the password
+  of an administrator
+- The "client IP" in a server-to-server communication is less sensitive than the
+  "client IP" in an client-to-server communication, where the client can be
+  linked to a human.
+- API tokens of a demo system are less sensitive than API tokens for a production
+  system
+- The license plate of an individual’s car is less sensitive than their social
+  security number
+- The full name of a user in a social network is less sensitive than the full
+  name of a user in a medical research database
+
+Depending on the sensitivity data an end-user of an observability system may
+weigh up if collecting this data is worth it.
+
+### Regulatory and other requirements
+
+Due to the negative effects that the exposure of sensitive data can have (see
+above in "Types of sensitive data"), different entities have created regulations
+for the collection of sensitive data, among them:
+
+- GDPR
+- CPRA
+- PIPEDA
+- HIPAA
+- [more…](https://en.wikipedia.org/wiki/Information_privacy)
+
+Additionally the entities operating the applications who leverage OpenTelemetry
+may have their own requirements for treating certain sensitive data.
+
+Finally end-users may want to apply recommendations for
+[Data Minimization](https://en.wikipedia.org/wiki/Data_minimization), to avoid
+"unnecessary risks for the data subject".
+
+**Note 1**: it is not (and can not be) the responsibility of the OpenTelemetry
+project to provide compliance with any of those regulations, this is a
+responsibility of the OTel end-user. OTel can only facilitate parts of those
+requirements.
+
+**Note 2**: Those requirements are subject of change and outside of the control
+of the OpenTelemetry community.
+
+## Trade-offs and mitigations
+
+### Performance Impact
+
+By adding an extra layer of processing every time an attribute value gets set,
+might have an impact on the performance. There might be ways to reduce that
+overhead, e.g. by only redacting values which are finalized and ready to exported
```

Comment:
This has security implications.  E.g. in Java we you wouldn't want sensitive passwords interned in Strings because they can be exfiltrated.

I like what you're suggesting, but I'd go the other way.  You could memoize the redaction function when generating the attribute value in some fashion, *or* use codegen/weaver to optimise these use cases (if we get the yaml model sorted).